modules/core/include/opencv2/core/cuda.hpp

```diff
@@ -337,6 +337,195 @@ class CV_EXPORTS GpuMat
     Allocator* allocator;
 };
 
+class CV_EXPORTS GpuMatND
+{
+public:
+    class CV_EXPORTS Allocator
+    {
+    public:
+        virtual ~Allocator() {}
+
+        // allocator must fill data, step and refcount fields
+        virtual bool allocate(GpuMatND* mat, int* shape, size_t elemSize) = 0;
+        virtual void free(GpuMatND* mat) = 0;
+    };
+
+    //! default allocator
+    static Allocator* defaultAllocator();
+    static void setDefaultAllocator(Allocator* allocator);
+
+    //! default constructor
+    explicit GpuMatND(Allocator* allocator = defaultAllocator());
+
+    //! constructs GpuMatND of the specified shape and type
+    GpuMatND(int* shape, int type, Allocator* allocator = defaultAllocator());
```

Comment:
We can use variable number of arguments as well here using the APIs in `cstdarg` but in other libraries like TF, the shape is passed as a list, so passing it as an array is better and avoids complications of variable number of arguments.

Reply:
> int* shape

This is unsafe API - missing `ndims`.

Consider targeting to "master" branch with C++11 support.
Use `std::array` and or `std::vector<int>` for passing ND array layouts.

Reply:
> Consider targeting to "master" branch with C++11 support.

Done.

> Use `std::array` and or `std::vector<int>` for passing ND array layouts.

I have used `std::vector<int>`.